presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,318 @@
+            else {
+                // If a future did not finish, or finished exceptionally, we do not add it to the results
+                // A new query for the same partition could trigger a successful quick stats fetch for this partition
+                result.put(partitionId, empty());
+            }
+        }
+
+        return result.build();
+    }
+
+    public PartitionStatistics getQuickStats(ConnectorSession session, SemiTransactionalHiveMetastore metastore, SchemaTableName table,
+            MetastoreContext metastoreContext, String partitionId)
+    {
+        if (!isQuickStatsEnabled(session)) {
+            return empty();
+        }
+        requestCount.incrementAndGet(); // New request was made to resolve quick stats for a partition
+
+        // Check if we already have stats cached in partitionIdToQuickStatsCache. If so return from cache
+        String partitionKey = String.join("/", table.toSchemaTablePrefix().toString(), partitionId);
+
+        PartitionStatistics cachedValue = partitionToStatsCache.getIfPresent(partitionKey);
+        if (cachedValue != null) {
+            succesfulResolveFromCacheCount.incrementAndGet();
+            return cachedValue;
+        }
+
+        // Check if we already have a quick stats build in progress for this partition key
+        if (inProgressBuilds.containsKey(partitionKey)) {
+            return empty();
+        }
+
+        // If not, atomically initiate a call to build quick stats in a background thread
+        AtomicReference<CompletableFuture<PartitionStatistics>> partitionStatisticsCompletableFuture = new AtomicReference<>();
+        inProgressBuilds.computeIfAbsent(partitionKey, (key) -> {
+            partitionStatisticsCompletableFuture.set(
+                    supplyAsync(() -> buildQuickStats(partitionKey, partitionId, session, metastore, table, metastoreContext), backgroundFetchExecutor));
+
+            // Also add a hook to reap this in-progress thread if it doesn't finish in reaperExpiry seconds
+            inProgressReaperExecutor.schedule(() -> {
+                inProgressBuilds.remove(partitionKey);
+                partitionStatisticsCompletableFuture.get().cancel(true);
+            }, reaperExpiryMillis, MILLISECONDS);
+            return Instant.now();
+        });
+
+        CompletableFuture<PartitionStatistics> future = partitionStatisticsCompletableFuture.get();
+        long quickStatsBuildTimeoutMillis = getQuickStatsBuildTimeoutMillis(session);
+
+        // A background call to build quick stats was started, and we want to wait for quick stats to be built
+        // Note : Only the first query that initiated the quick stats call for this partition will have to wait for the stats to be built
+        // Other queries will either observe an in-progress build, and get an empty quick stats returned from below
+        // Or would've already gotten the cached quick stats value
+        if (future != null && quickStatsBuildTimeoutMillis > 0) {
```

Comment:
Do we know how this behaves in case of a self-join? I hope we do not wind up split brained and use stats for partitions in one part of the query and not in another. Maybe another good test case

Reply:
Since the planner code is single threaded for an individual query, stats are resolved one PlanNode at a time. So the first call to resolve quick stats will come from the first instance of the TableScan for this table, this will resolve and cache stats. The next call from the subsequent TableScan node will pick up these cached stats.

I could confirm this with a query : `explain (format json) select 1 from region_no_stats r1, region_no_stats r2 where r1.r_regionkey = r2.r_regionkey;` , [the results](https://gist.github.com/aaneja/680aea3681ef7d592f0e20b8dbb92776#file-quickstats_selfjoin-json) show that the stats for each TableScan node were correctly resolved to -
```
 "r_regionkey<integer>" : {                                                                                                                                                                                                                                                          
                 "lowValue" : 0.0,                                                                                                                                                                                                                                                                 
                 "highValue" : 4.0,                                                                                                                                                                                                                                                                
                 "nullsFraction" : 0.0,                                                                                                                                                                                                                                                            
                 "averageRowSize" : "NaN",                                                                                                                                                                                                                                                         
                 "distinctValuesCount" : "NaN"                                                                                                                                                                                                                                                     
               } 
```

Now if the first call came in and the quick stats resolution times out, this node will see `EMPTY` stats. The next call could, in theory, see the cached stats if the quick stats resolution finishes in the background.

While this is not ideal, I don't see this as incorrect behavior - the same behavior would occur with HMS based stats too; a background HMS update whilst a query is in progress could in theory result in the same situation